File: services/golang/internal/handler/task_handler.go
package handler

import (
	"context"
	"encoding/json"

	"connectrpc.com/connect"
	"github.com/google/uuid"
	"google.golang.org/protobuf/types/known/structpb"
	"google.golang.org/protobuf/types/known/timestamppb"

	commonv1 "github.com/igorrmotta/api-corestack/services/golang/gen/common/v1"
	taskv1 "github.com/igorrmotta/api-corestack/services/golang/gen/task/v1"
	"github.com/igorrmotta/api-corestack/services/golang/gen/task/v1/taskv1connect"
	"github.com/igorrmotta/api-corestack/services/golang/internal/domain"
	"github.com/igorrmotta/api-corestack/services/golang/internal/service"
)

type TaskHandler struct {
	taskv1connect.UnimplementedTaskServiceHandler
	svc       *service.TaskService
	importSvc *service.ImportService
}

func NewTaskHandler(svc *service.TaskService, importSvc *service.ImportService) *TaskHandler {
	return &TaskHandler{svc: svc, importSvc: importSvc}
}

func (h *TaskHandler) CreateTask(ctx context.Context, req *connect.Request[taskv1.CreateTaskRequest]) (*connect.Response[taskv1.CreateTaskResponse], error) {
	workspaceID, err := uuid.Parse(req.Msg.WorkspaceId)
	if err != nil {
		return nil, connect.NewError(connect.CodeInvalidArgument, err)
	}
	projectID, err := uuid.Parse(req.Msg.ProjectId)
	if err != nil {
		return nil, connect.NewError(connect.CodeInvalidArgument, err)
	}

	params := domain.CreateTaskParams{
		WorkspaceID: workspaceID,
		ProjectID:   projectID,
		Title:       req.Msg.Title,
		Description: req.Msg.Description,
		Priority:    req.Msg.Priority,
		AssignedTo:  req.Msg.AssignedTo,
	}
	if req.Msg.DueDate != nil {
		t := req.Msg.DueDate.AsTime()
		params.DueDate = &t
	}
	if req.Msg.Metadata != nil {
		b, err := json.Marshal(req.Msg.Metadata.AsMap())
		if err != nil {
			return nil, connect.NewError(connect.CodeInvalidArgument, err)
		}
		params.Metadata = b
	}

	task, err := h.svc.Create(ctx, params)
	if err != nil {
		return nil, toConnectError(err)
	}
	proto, err := taskToProto(task)
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}
	return connect.NewResponse(&taskv1.CreateTaskResponse{
		Task: proto,
	}), nil
}

func (h *TaskHandler) GetTask(ctx context.Context, req *connect.Request[taskv1.GetTaskRequest]) (*connect.Response[taskv1.GetTaskResponse], error) {
	id, err := uuid.Parse(req.Msg.Id)
	if err != nil {
		return nil, connect.NewError(connect.CodeInvalidArgument, err)
	}
	task, err := h.svc.GetByID(ctx, id)
	if err != nil {
		return nil, toConnectError(err)
	}
	proto, err := taskToProto(task)
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}
	return connect.NewResponse(&taskv1.GetTaskResponse{
		Task: proto,
	}), nil
}

func (h *TaskHandler) ListTasks(ctx context.Context, req *connect.Request[taskv1.ListTasksRequest]) (*connect.Response[taskv1.ListTasksResponse], error) {
	workspaceID, err := uuid.Parse(req.Msg.WorkspaceId)
	if err != nil {
		return nil, connect.NewError(connect.CodeInvalidArgument, err)
	}

	params := domain.ListTasksParams{
		WorkspaceID: workspaceID,
		Status:      req.Msg.Status,
		Priority:    req.Msg.Priority,
		AssignedTo:  req.Msg.AssignedTo,
	}
	if req.Msg.ProjectId != "" {
		projectID, err := uuid.Parse(req.Msg.ProjectId)
		if err != nil {
			return nil, connect.NewError(connect.CodeInvalidArgument, err)
		}
		params.ProjectID = projectID
	}
	if req.Msg.Pagination != nil {
		params.PageSize = req.Msg.Pagination.PageSize
		params.PageToken = req.Msg.Pagination.PageToken
	}

	list, err := h.svc.List(ctx, params)
	if err != nil {
		return nil, toConnectError(err)
	}
	tasks := make([]*taskv1.Task, len(list.Tasks))
	for i, t := range list.Tasks {
		proto, err := taskToProto(&t)
		if err != nil {
			return nil, connect.NewError(connect.CodeInternal, err)
		}
		tasks[i] = proto
	}
	return connect.NewResponse(&taskv1.ListTasksResponse{
		Tasks: tasks,
		Pagination: &commonv1.PaginationResponse{
			NextPageToken: list.NextPageToken,
			TotalCount:    list.TotalCount,
		},
	}), nil
}

func (h *TaskHandler) UpdateTask(ctx context.Context, req *connect.Request[taskv1.UpdateTaskRequest]) (*connect.Response[taskv1.UpdateTaskResponse], error) {
	id, err := uuid.Parse(req.Msg.Id)
	if err != nil {
		return nil, connect.NewError(connect.CodeInvalidArgument, err)
	}

	params := domain.UpdateTaskParams{
		ID:          id,
		Title:       req.Msg.Title,
		Description: req.Msg.Description,
		Status:      req.Msg.Status,
		Priority:    req.Msg.Priority,
		AssignedTo:  req.Msg.AssignedTo,
	}
	if req.Msg.DueDate != nil {
		t := req.Msg.DueDate.AsTime()
		params.DueDate = &t
	}
	if req.Msg.Metadata != nil {
		b, err := json.Marshal(req.Msg.Metadata.AsMap())
		if err != nil {
			return nil, connect.NewError(connect.CodeInvalidArgument, err)
		}
		params.Metadata = b
	}

	task, err := h.svc.Update(ctx, params)
	if err != nil {
		return nil, toConnectError(err)
	}
	proto, err := taskToProto(task)
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}
	return connect.NewResponse(&taskv1.UpdateTaskResponse{
		Task: proto,
	}), nil
}

func (h *TaskHandler) DeleteTask(ctx context.Context, req *connect.Request[taskv1.DeleteTaskRequest]) (*connect.Response[taskv1.DeleteTaskResponse], error) {
	id, err := uuid.Parse(req.Msg.Id)
	if err != nil {
		return nil, connect.NewError(connect.CodeInvalidArgument, err)
	}
	if err := h.svc.Delete(ctx, id); err != nil {
		return nil, toConnectError(err)
	}
	return connect.NewResponse(&taskv1.DeleteTaskResponse{}), nil
}

func (h *TaskHandler) BulkImportTasks(ctx context.Context, req *connect.Request[taskv1.BulkImportTasksRequest]) (*connect.Response[taskv1.BulkImportTasksResponse], error) {
	workspaceID, err := uuid.Parse(req.Msg.WorkspaceId)
	if err != nil {
		return nil, connect.NewError(connect.CodeInvalidArgument, err)
	}
	projectID, err := uuid.Parse(req.Msg.ProjectId)
	if err != nil {
		return nil, connect.NewError(connect.CodeInvalidArgument, err)
	}

	inputs := make([]domain.TaskInput, len(req.Msg.Tasks))
	for i, t := range req.Msg.Tasks {
		input := domain.TaskInput{
			Title:       t.Title,
			Description: t.Description,
			Priority:    t.Priority,
			AssignedTo:  t.AssignedTo,
		}
		if t.DueDate != nil {
			d := t.DueDate.AsTime()
			input.DueDate = &d
		}
		if t.Metadata != nil {
			b, err := json.Marshal(t.Metadata.AsMap())
			if err != nil {
				return nil, connect.NewError(connect.CodeInvalidArgument, err)
			}
			input.Metadata = b
		}
		inputs[i] = input
	}

	result := h.importSvc.BulkImport(ctx, workspaceID, projectID, inputs)

	taskErrors := make([]*taskv1.TaskError, len(result.Errors))
	for i, e := range result.Errors {
		taskErrors[i] = &taskv1.TaskError{
			Index: e.Index,
			Error: e.Error,
		}
	}

	return connect.NewResponse(&taskv1.BulkImportTasksResponse{
		Total:     result.Total,
		Succeeded: result.Succeeded,
		Failed:    result.Failed,
		Errors:    taskErrors,
	}), nil
}

func taskToProto(t *domain.Task) (*taskv1.Task, error) {
	proto := &taskv1.Task{
		Id:          t.ID.String(),
		WorkspaceId: t.WorkspaceID.String(),
		ProjectId:   t.ProjectID.String(),
		Title:       t.Title,
		Description: t.Description,
		Status:      t.Status,
		Priority:    t.Priority,
		AssignedTo:  t.AssignedTo,
		CreatedAt:   timestamppb.New(t.CreatedAt),
		UpdatedAt:   timestamppb.New(t.UpdatedAt),
	}
	if t.DueDate != nil {
		proto.DueDate = timestamppb.New(*t.DueDate)
	}
	if len(t.Metadata) > 0 {
		var m map[string]any
		if err := json.Unmarshal(t.Metadata, &m); err != nil {
			return nil, err
		}
		s, err := structpb.NewStruct(m)
		if err != nil {
			return nil, err
		}
		proto.Metadata = s
	}
	return proto, nil
}

File: services/typescript/src/handler/task-handler.ts
import type { ConnectRouter } from "@connectrpc/connect";
import { ConnectError, Code } from "@connectrpc/connect";
import { TaskService as TaskServiceDef } from "../../gen/task/v1/task_pb.js";
import { TaskService } from "../service/task-service.js";
import { ImportService } from "../service/import-service.js";
import { toConnectError } from "./error-mapping.js";
import { timestampFromDate, timestampDate } from "@bufbuild/protobuf/wkt";
import type { JsonObject } from "@bufbuild/protobuf";
import type { Task } from "../repository/task-repo.js";

function taskToProto(t: Task) {
  const proto: Record<string, unknown> = {
    id: t.id,
    workspaceId: t.workspaceId,
    projectId: t.projectId,
    title: t.title,
    description: t.description,
    status: t.status,
    priority: t.priority,
    assignedTo: t.assignedTo,
    createdAt: timestampFromDate(t.createdAt),
    updatedAt: timestampFromDate(t.updatedAt),
  };
  if (t.dueDate) {
    proto.dueDate = timestampFromDate(t.dueDate);
  }
  if (t.metadata && Object.keys(t.metadata).length > 0) {
    proto.metadata = t.metadata as JsonObject;
  }
  return proto;
}

export function registerTaskHandler(
  router: ConnectRouter,
  svc: TaskService,
  importSvc: ImportService,
) {
  router.service(TaskServiceDef, {
    async createTask(req) {
      if (!req.workspaceId) {
        throw new ConnectError("workspace_id is required", Code.InvalidArgument);
      }
      if (!req.projectId) {
        throw new ConnectError("project_id is required", Code.InvalidArgument);
      }
      try {
        const task = await svc.create({
          workspaceId: req.workspaceId,
          projectId: req.projectId,
          title: req.title,
          description: req.description,
          priority: req.priority,
          assignedTo: req.assignedTo,
          dueDate: req.dueDate ? timestampDate(req.dueDate) : null,
          metadata: (req.metadata as Record<string, unknown>) ?? null,
        });
        return { task: taskToProto(task) };
      } catch (err) {
        throw toConnectError(err);
      }
    },

    async getTask(req) {
      if (!req.id) {
        throw new ConnectError("id is required", Code.InvalidArgument);
      }
      try {
        const task = await svc.getById(req.id);
        return { task: taskToProto(task) };
      } catch (err) {
        throw toConnectError(err);
      }
    },

    async listTasks(req) {
      if (!req.workspaceId) {
        throw new ConnectError("workspace_id is required", Code.InvalidArgument);
      }
      try {
        const list = await svc.list({
          workspaceId: req.workspaceId,
          projectId: req.projectId,
          status: req.status,
          priority: req.priority,
          assignedTo: req.assignedTo,
          pageSize: req.pagination?.pageSize ?? 0,
          pageToken: req.pagination?.pageToken ?? "",
        });
        return {
          tasks: list.tasks.map(taskToProto),
          pagination: {
            nextPageToken: list.nextPageToken,
            totalCount: list.totalCount,
          },
        };
      } catch (err) {
        throw toConnectError(err);
      }
    },

    async updateTask(req) {
      if (!req.id) {
        throw new ConnectError("id is required", Code.InvalidArgument);
      }
      try {
        const task = await svc.update({
          id: req.id,
          title: req.title,
          description: req.description,
          status: req.status,
          priority: req.priority,
          assignedTo: req.assignedTo,
          dueDate: req.dueDate ? timestampDate(req.dueDate) : null,
          metadata: (req.metadata as Record<string, unknown>) ?? null,
        });
        return { task: taskToProto(task) };
      } catch (err) {
        throw toConnectError(err);
      }
    },

    async deleteTask(req) {
      if (!req.id) {
        throw new ConnectError("id is required", Code.InvalidArgument);
      }
      try {
        await svc.delete(req.id);
        return {};
      } catch (err) {
        throw toConnectError(err);
      }
    },

    async bulkImportTasks(req) {
      if (!req.workspaceId) {
        throw new ConnectError("workspace_id is required", Code.InvalidArgument);
      }
      if (!req.projectId) {
        throw new ConnectError("project_id is required", Code.InvalidArgument);
      }
      try {
        const inputs = req.tasks.map((t) => ({
          title: t.title,
          description: t.description,
          priority: t.priority,
          assignedTo: t.assignedTo,
          dueDate: t.dueDate ? timestampDate(t.dueDate) : null,
          metadata: (t.metadata as Record<string, unknown>) ?? null,
        }));

        const result = await importSvc.bulkImport(req.workspaceId, req.projectId, inputs);

        return {
          total: result.total,
          succeeded: result.succeeded,
          failed: result.failed,
          errors: result.errors.map((e) => ({
            index: e.index,
            error: e.error,
          })),
        };
      } catch (err) {
        throw toConnectError(err);
      }
    },
  });
}
